### PR TITLE
[WIP] OPS-1379 As an operator I want automation to deploy API gateways

### PR DESCRIPTION
**Pull request by arbabnazar**
@e0d can you please review the progress. Thanks

### CODE REVIEW

File: terraform/aws-api-gateway/variables.tf [hunk 1]
```diff
@@ -0,0 +1,35 @@
+variable rds_password {
```
Comment (e0d) [inline terraform/aws-api-gateway/variables.tf]: cut and paste?

File: terraform/aws-api-gateway/main.tf [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+resource "aws_api_gateway_integration_response" "IntegrationRequest" {
+  rest_api_id = "${aws_api_gateway_rest_api.api.id}"
+  resource_id = "${aws_api_gateway_resource.res.id}"
+  http_method = "${aws_api_gateway_method.resource_method.http_method}"
+  status_code = "${aws_api_gateway_method_response.MethodResponse.status_code}"
+  selection_pattern = "${var.api_selection_pattern}"
+}
+
+resource "aws_api_gateway_deployment" "test" {
```
Comment (e0d) [inline terraform/aws-api-gateway/main.tf]: what is "test" here?

File: terraform/aws-api-gateway/main.tf [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+resource "aws_api_gateway_method" "resource_method" {
+  rest_api_id = "${aws_api_gateway_rest_api.api.id}"
+  resource_id = "${aws_api_gateway_resource.res.id}"
+  http_method = "${var.http_method}"
+  authorization = "${var.authorization}"
+}
+
+resource "aws_api_gateway_integration" "MethodRequest" {
```
Comment (e0d) [inline terraform/aws-api-gateway/main.tf]: there's a mixture of snake case foo_bar and standing camel case, FooBar in the resource name, let's standardize and follow the common pattern.

**Comment by arbabnazar** [comment]
@e0d please check the progress on cloudfront and route53. Thanks

File: terraform/cloudfront/main.tf [hunk 4]
```diff
@@ -0,0 +1,74 @@
+provider "aws" {
+    region = "${var.aws_region}"
+}
+
+// Setup your S3 Bucket
+resource "aws_s3_bucket" "cdn_bucket" {
```
Comment (e0d) [inline terraform/cloudfront/main.tf:6]: Unless we need to specify some origin when creating the distribution, this should not be necessary.  The origin in our case will be the API gateway itself.

File: terraform/cloudfront/main.tf [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+// Setup the CloudFront Distribution
+resource "aws_cloudfront_distribution" "cloudfront_distribution" {
+  origin {
+    domain_name = "${var.bucket_name}.s3.amazonaws.com"
+    origin_id = "S3-${var.bucket_name}"
+    s3_origin_config {}
+  }
+  enabled = true
+  //aliases = ["mytestapi.com", "www.mytestapi.com"]
```
Comment (e0d) [inline terraform/cloudfront/main.tf:34]: remove commented code.

File: terraform/cloudfront/main.tf [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  price_class = "${var.price_class}"
+  default_cache_behavior {
+    allowed_methods = [ "DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT" ]
+    cached_methods = [ "GET", "HEAD" ]
+    target_origin_id = "S3-${var.bucket_name}"
```
Comment (e0d) [inline terraform/cloudfront/main.tf:39]: This should by the API gateway.

File: terraform/cloudfront/main.tf [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    forwarded_values {
+      query_string = true
+      cookies {
+        forward = "none"
+      }
+    }
+    viewer_protocol_policy = "allow-all"
```
Comment (e0d) [inline terraform/cloudfront/main.tf:46]: @efagin do we want to terminate SSL at CF?  Whose responsibility should wire-encryption be?

File: terraform/cloudfront/main.tf [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    min_ttl = 0
```
Comment (e0d) [inline terraform/cloudfront/main.tf:47]: The TTLs should be vars and we'll want to set the min TTL to something other than 0.  For now, let's have it match the default_ttl.

File: terraform/cloudfront/main.tf [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  //aliases = ["mytestapi.com", "www.mytestapi.com"]
+  price_class = "${var.price_class}"
+  default_cache_behavior {
+    allowed_methods = [ "DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT" ]
+    cached_methods = [ "GET", "HEAD" ]
+    target_origin_id = "S3-${var.bucket_name}"
+    forwarded_values {
+      query_string = true
+      cookies {
+        forward = "none"
```
Comment (e0d) [inline terraform/cloudfront/main.tf:43]: @efagin what's the appropriate cookie forwarding policy?

File: terraform/cloudfront/main.tf (hunk 7 above)
Comment (efagin) [inline terraform/cloudfront/main.tf:46]: @e0d no, SSL termination is still the responsibility of the origin. Also - why all the CloudFront code? API Gateway manages an internal CF distribution but that isn't exposed to us and we certainly don't have any control over it. You shouldn't have to touch CF to manage an API gateway - the only non-APIG service you should be messing with is Route53 for the one-time CNAME to the gateway-generated distribution ID. Let me know if you guys want a walkthrough.

File: terraform/aws-api-gateway/main.tf [hunk 10]
```diff
@@ -0,0 +1,55 @@
+provider "aws" {
+    region = "${var.aws_region}"
+}
+
+resource "aws_api_gateway_rest_api" "api" {
+  name = "${var.api_name}"
+  description = "${var.api_description}"
+}
+
+resource "aws_api_gateway_resource" "res" {
```
Comment (efagin) [inline terraform/aws-api-gateway/main.tf:10]: I don't understand why we care about resources, methods, or integrations at all? As far as we're concerned these are API Gateway internals, provided we're using the ImportRestAPI.

File: terraform/cloudfront/main.tf (hunk 7 above)
Comment (e0d) [inline terraform/cloudfront/main.tf:46]: When we groomed this we added an AC to create a CF distribution.  Was not clear this was automatic, it was clear the gateway controlled some internal details.  We can remove this code.

File: terraform/cloudfront/main.tf (hunk 9 above)
Comment (efagin) [inline terraform/cloudfront/main.tf:43]: Again, not something CF should be managing at all. Individual origins can specify in their Swagger fragments if they want the Cookie header passed through (in full) or not. Individual cookie filters aren't supported, at least at the moment, it's the entire header value.

File: terraform/cloudfront/main.tf (hunk 7 above)
Comment (efagin) [inline terraform/cloudfront/main.tf:46]: BTW "origin" generally means some service-specific ELB, so what I mean to say is SSL termination is probably happening at the ELB layer (or lower) in general, and this gateway shouldn't change that policy. Especially if we have multiple gateways or access pathways.

**Comment by efagin** [comment]
I honestly don't understand why the majority of this pull request is necessary - to summarize some of my inline comments, API gateway automatically manages its private CloudFront distribution and low-level objects like methods and integrations, and exposes higher-level APIs for us (ImportRestApi and CreateDomainName specifically) to manage those resources.

**Pull request closed without merging**